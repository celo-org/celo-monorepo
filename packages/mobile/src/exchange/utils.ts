import BigNumber from 'bignumber.js'
import { CURRENCY_ENUM } from 'src/geth/consts'
import { convertToContractDecimals } from 'src/tokens/saga'
import { getContractKitAsync } from 'src/web3/contracts'

export async function celoTocUsd(celoAmount: BigNumber) {
  const celoAmountInWei = (
    await convertToContractDecimals(celoAmount, CURRENCY_ENUM.GOLD)
  ).integerValue()

  const contractKit = await getContractKitAsync()
  const exchange = await contractKit.contracts.getExchange()
  const exchangeRate = await exchange.getGoldExchangeRate(celoAmountInWei)
  const cUsdAmount = exchangeRate.multipliedBy(celoAmount)

  return cUsdAmount
}
