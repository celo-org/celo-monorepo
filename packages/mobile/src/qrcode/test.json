{
    "address": "0xf7f551752A78Ce650385B58364225e5ec18D96cB",
    "displayName": "Super 8",
    "currencyCode": "PHP",
    "amount": "500",
    "comment": "92a53156-c0f2-11ea-b3de-0242ac13000"
}