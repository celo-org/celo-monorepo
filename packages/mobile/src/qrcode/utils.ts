import { isValidAddress } from '@celo/utils/src/address'
import { isEmpty } from 'lodash'
import * as RNFS from 'react-native-fs'
import Share from 'react-native-share'
import { put } from 'redux-saga/effects'
import { showError } from 'src/alert/actions'
import { ErrorMessages } from 'src/app/ErrorMessages'
import { AddressToE164NumberType } from 'src/identity/reducer'
import { navigate } from 'src/navigator/NavigationService'
import { Screens } from 'src/navigator/Screens'
import {
  getRecipientFromAddress,
  NumberToRecipient,
  Recipient,
  RecipientKind,
} from 'src/recipients/recipient'
import { QrCode, storeLatestInRecents, SVG } from 'src/send/actions'
import Logger from 'src/utils/Logger'

export enum BarcodeTypes {
  QR_CODE = 'QR_CODE',
}

const TAG = 'QR/utils'

const QRFileName = '/celo-qr.png'

export async function shareSVGImage(svg: SVG) {
  if (!svg) {
    return
  }
  svg.toDataURL(async (data: string) => {
    const path = RNFS.DocumentDirectoryPath + QRFileName
    try {
      await RNFS.writeFile(path, data, 'base64')
      Share.open({
        url: 'file://' + path,
        type: 'image/png',
      }).catch((err: Error) => {
        throw err
      })
    } catch (e) {
      Logger.warn(TAG, e)
    }
  })
}

export function* handleBarcode(
  barcode: QrCode,
  addressToE164Number: AddressToE164NumberType,
  recipientCache: NumberToRecipient
) {
  let data: { address: string; e164PhoneNumber: string; displayName: string } | undefined
  try {
    data = JSON.parse(barcode.data)
  } catch (e) {
    Logger.warn(TAG, 'QR code read failed with ' + e)
  }
  if (typeof data !== 'object' || isEmpty(data.address)) {
    yield put(showError(ErrorMessages.QR_FAILED_NO_ADDRESS))
    return
  }
  if (!isValidAddress(data.address)) {
    yield put(showError(ErrorMessages.QR_FAILED_INVALID_ADDRESS))
    return
  }
  if (typeof data.e164PhoneNumber !== 'string') {
    // Default for invalid e164PhoneNumber
    data.e164PhoneNumber = ''
  }
  if (typeof data.displayName !== 'string') {
    // Default for invalid displayName
    data.displayName = ''
  }
  const cachedRecipient = getRecipientFromAddress(data.address, addressToE164Number, recipientCache)

  const recipient: Recipient = cachedRecipient
    ? {
        ...data,
        kind: RecipientKind.QrCode,
        displayId: data.e164PhoneNumber,
        phoneNumberLabel: 'QR Code',
        thumbnailPath: cachedRecipient.thumbnailPath,
        contactId: cachedRecipient.contactId,
      }
    : {
        ...data,
        kind: RecipientKind.QrCode,
        displayId: data.e164PhoneNumber,
      }
  yield put(storeLatestInRecents(recipient))

  navigate(Screens.SendAmount, { recipient })
}
