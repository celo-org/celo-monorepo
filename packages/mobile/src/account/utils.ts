import { ADDRESS_LENGTH } from 'src/exchange/reducer'

export const isAddressFormat = (content: string): boolean => {
  return content.startsWith('0x') && content.length === ADDRESS_LENGTH
}
