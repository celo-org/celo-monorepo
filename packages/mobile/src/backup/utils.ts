import { sampleSize } from 'lodash'
import { AsyncStorage } from 'react-native'
import { generateMnemonic, wordlists } from 'react-native-bip39'
import { getKey, setKey } from 'src/utils/keyStore'
import Logger from 'src/utils/Logger'

const TAG = 'Backup/utils'

export async function createQuizWordList(mnemonic: string, language: string | null) {
  const disallowedWordSet = new Set(mnemonic.split(' '))
  const languageWordList = getWordlist(language)
  const wordOptions: string = await generateMnemonic(1000, null, languageWordList)
  const quizWordList = new Set(
    [...wordOptions.split(' ')].filter((word: string) => !disallowedWordSet.has(word))
  )
  return [...quizWordList]
}

export function selectQuizWordOptions(correctWord: string, allWords: string[], numOptions: number) {
  const wordOptions = []
  const correctWordPosition = Math.floor(Math.random() * numOptions)
  const randomWordIndexList = sampleSize([...Array(allWords.length).keys()], numOptions - 1)
  let randomWordIndex: number = 0

  for (let i = 0; i < numOptions; i++) {
    if (i === correctWordPosition) {
      wordOptions.push(correctWord)
      continue
    }

    wordOptions.push(allWords[randomWordIndexList[randomWordIndex]])
    randomWordIndex += 1
  }
  return wordOptions
}

export function getWordlist(language: string | null) {
  if (!language) {
    return wordlists.EN
  }

  switch (language.slice(0, 2)) {
    case 'es': {
      return wordlists.ES
    }
    default: {
      return wordlists.EN
    }
  }
}

// TODO(Rossy) Remove after the next alfa testnet reset
export async function getStoredMnemonic(): Promise<string | null> {
  try {
    Logger.debug(TAG, 'Checking keystore for mnemonic')
    let mnemonic = await getKey('mnemonic')
    if (mnemonic) {
      return mnemonic
    }

    Logger.debug(TAG, 'Mnemonic not found in keystore, checking async storage')
    mnemonic = await AsyncStorage.getItem('mnemonic')
    if (mnemonic) {
      await setKey('mnemonic', mnemonic)
      await AsyncStorage.removeItem('mnemonic')
      return mnemonic
    }

    Logger.error(TAG, 'No mnemonic found')
    return null
  } catch (error) {
    Logger.error(TAG, 'Failed to retrieve mnemonic', error)
    return null
  }
}
