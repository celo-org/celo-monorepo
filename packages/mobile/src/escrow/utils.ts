import { ensureLeading0x } from '@celo/base'
import { generateDeterministicInviteCode } from '@celo/utils/lib/account'
import { publicKeyToAddress } from '@celo/utils/lib/address'

export const generateEscrowPaymentIdAndPk = (
  recipientPhoneHash: string,
  recipientPepper: string,
  addressIndex: number = 0
) => {
  const { publicKey, privateKey } = generateDeterministicInviteCode(
    recipientPhoneHash,
    recipientPepper,
    addressIndex
  )
  return { paymentId: publicKeyToAddress(publicKey), privateKey: ensureLeading0x(privateKey) }
}

export const generateUniquePaymentId = (
  existingPaymentIds: string[],
  phoneHash: string,
  pepper: string
) => {
  const paymentIdSet: Set<string> = new Set(existingPaymentIds)

  // Using an upper bound of 100 to be sure this doesn't run forever given
  // the realistic amount of pending escrow txs is far less than this
  for (let i = 0; i < 100; i += 1) {
    const { paymentId } = generateEscrowPaymentIdAndPk(phoneHash, pepper, i)
    if (!paymentIdSet.has(paymentId)) {
      return paymentId
    }
  }
}
