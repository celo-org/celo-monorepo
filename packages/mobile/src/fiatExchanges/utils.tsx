import { SIMPLEX_URI } from 'src/config'
import { CURRENCY_ENUM } from 'src/geth/consts'
import { LocalCurrencyCode } from 'src/localCurrency/consts'
import { navigate } from 'src/navigator/NavigationService'
import { Screens } from 'src/navigator/Screens'
import { navigateToURI } from 'src/utils/linking'

export const openMoonpay = (currencyCode: LocalCurrencyCode, currencyToBuy: CURRENCY_ENUM) => {
  navigate(Screens.MoonPay, {
    localAmount: 0,
    currencyCode,
    currencyToBuy,
  })
}

export const openSimplex = (account: string | null) => {
  navigateToURI(`${SIMPLEX_URI}?address=${account}`)
}
