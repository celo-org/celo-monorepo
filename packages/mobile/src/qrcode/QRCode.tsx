import Button, { BtnTypes } from '@celo/react-components/components/Button'
import DownloadIcon from '@celo/react-components/icons/Download'
import ScanCodeIcon from '@celo/react-components/icons/ScanCode'
import colors from '@celo/react-components/styles/colors'
import { fontStyles } from '@celo/react-components/styles/fonts'
import variables from '@celo/react-components/styles/variables'
import * as React from 'react'
import { WithNamespaces, withNamespaces } from 'react-i18next'
import { StyleSheet, View } from 'react-native'
import QRCode from 'react-native-qrcode-svg'
import { connect } from 'react-redux'
import { UserContactDetails, userContactDetailsSelector } from 'src/account/reducer'
import { AvatarSelf } from 'src/components/AvatarSelf'
import i18n, { Namespaces } from 'src/i18n'
import { headerWithBackButton } from 'src/navigator/Headers'
import { navigate } from 'src/navigator/NavigationService'
import { Screens } from 'src/navigator/Screens'
import { RootState } from 'src/redux/reducers'
import { shareQRCode, SVG } from 'src/send/actions'
import { currentAccountSelector } from 'src/web3/selectors'

interface DispatchProps {
  shareQRCode: typeof shareQRCode
}

interface State {
  qrContent: string
}

interface StateProps {
  name: string
  e164Number: string
  userContact: UserContactDetails
  account: string | null
  defaultCountryCode: string
}

type Props = StateProps & DispatchProps & WithNamespaces

const mapStateToProps = (state: RootState) => {
  return {
    name: state.account.name,
    e164Number: state.account.e164PhoneNumber,
    userContact: userContactDetailsSelector(state),
    account: currentAccountSelector(state),
    defaultCountryCode: state.account.defaultCountryCode,
  }
}

const mapDispatchToProps = {
  shareQRCode,
}

class QRCodeDisplay extends React.Component<Props, State> {
  static navigationOptions = () => ({
    ...headerWithBackButton,
    headerTitle: i18n.t('sendFlow7:QRCode'),
  })

  state = {
    qrContent: '{}',
  }

  svg: SVG | null = null

  qrCodeRefReady = (c: SVG) => (this.svg = c)

  onPressShare = () => this.props.shareQRCode(this.svg)

  onPressScanCode = () => {
    navigate(Screens.QRScanner)
  }

  componentDidMount() {
    const { name, account, e164Number } = this.props
    const qrContent = JSON.stringify({
      address: account,
      e164PhoneNumber: e164Number,
      displayName: name,
    })
    this.setState({ qrContent })
  }

  render() {
    const { t } = this.props

    return (
      <View style={style.container}>
        <View style={style.userInfo}>
          <AvatarSelf />
        </View>
        <QRCode
          value={this.state.qrContent}
          size={variables.width / 2}
          color={colors.darkGrey}
          getRef={this.qrCodeRefReady}
        />
        <View style={style.buttonContainer}>
          <Button
            onPress={this.onPressShare}
            text={t('shareCodeImage')}
            accessibilityLabel={t('shareCodeImage')}
            standard={true}
            type={BtnTypes.PRIMARY}
            disabled={false}
            style={[fontStyles.buttonText]}
          >
            <DownloadIcon />
          </Button>
          <Button
            onPress={this.onPressScanCode}
            text={t('scanCode')}
            accessibilityLabel={t('scanCode')}
            standard={true}
            type={BtnTypes.SECONDARY}
            disabled={false}
          >
            <ScanCodeIcon color={colors.celoGreen} />
          </Button>
        </View>
      </View>
    )
  }
}

const style = StyleSheet.create({
  container: {
    flex: 1,
    alignItems: 'center',
    justifyContent: 'space-between',
    backgroundColor: colors.background,
    height: variables.height,
  },
  userInfo: {
    marginVertical: 20,
  },
  nameText: {
    fontSize: 16,
    fontFamily: 'Hindi Siliguri',
    fontWeight: '500',
    color: colors.dark,
    paddingTop: 20,
    paddingBottom: 5,
    textAlign: 'center',
  },
  buttonContainer: {
    width: variables.width,
    paddingHorizontal: 20,
    flex: 1,
    justifyContent: 'flex-end',
  },
  buttonScan: {
    color: colors.celoGreen,
  },
})

export default connect<StateProps, DispatchProps, {}, RootState>(
  mapStateToProps,
  mapDispatchToProps
)(withNamespaces(Namespaces.sendFlow7)(QRCodeDisplay))
