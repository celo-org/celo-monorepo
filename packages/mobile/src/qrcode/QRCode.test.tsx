import * as React from 'react'
import 'react-native'
import { Provider } from 'react-redux'
import * as renderer from 'react-test-renderer'
import QRCode from 'src/qrcode/QRCode'
import { createMockStore } from 'test/utils'
import { mockAccount, mockName } from 'test/values'

describe('QRCode', () => {
  const store = createMockStore({
    account: { name: mockName },
    web3: {
      account: mockAccount,
    },
  })
  it('renders correctly', () => {
    const tree = renderer.create(
      <Provider store={store}>
        <QRCode />
      </Provider>
    )

    expect(tree).toMatchSnapshot()
  })
})
