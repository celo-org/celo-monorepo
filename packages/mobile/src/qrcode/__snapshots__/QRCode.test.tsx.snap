// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QRCode renders correctly 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#FFFFFF",
      "flex": 1,
      "height": 1292,
      "justifyContent": "space-between",
    }
  }
>
  <View
    style={
      Object {
        "marginVertical": 20,
      }
    }
  >
    <View
      style={
        Object {
          "flexDirection": "column",
          "justifyContent": "center",
          "marginBottom": 10,
        }
      }
    >
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "flexDirection": "row",
              "justifyContent": "center",
            },
            Object {
              "alignSelf": "center",
              "margin": "auto",
            },
          ]
        }
      >
        <View
          style={
            Array [
              Object {
                "alignItems": "center",
                "justifyContent": "center",
              },
              Object {
                "backgroundColor": "#7AD6FE",
                "borderRadius": 20,
                "height": 40,
                "width": 40,
              },
            ]
          }
        >
          <Image
            source={
              Object {
                "testUri": "../../../src/images/unknown-user-icon.png",
              }
            }
            style={
              Array [
                Object {
                  "alignSelf": "center",
                  "margin": "auto",
                },
                Object {
                  "height": 40,
                  "width": 40,
                },
              ]
            }
          />
        </View>
      </View>
      <Text
        ellipsizeMode="tail"
        numberOfLines={1}
        style={
          Array [
            Object {
              "color": "#2E3338",
              "fontFamily": "Hind-SemiBold",
              "fontSize": 14,
              "lineHeight": 18,
            },
            Object {
              "marginHorizontal": 20,
              "paddingTop": 6,
              "textAlign": "center",
            },
          ]
        }
      >
        John Doe
      </Text>
    </View>
  </View>
  <QRCode
    color="#3D3D3D"
    getRef={[Function]}
    size={375}
    value="{\\"address\\":\\"0x0000000000000000000000000000000000007E57\\",\\"e164PhoneNumber\\":\\"+14155556666\\",\\"displayName\\":\\"John Doe\\"}"
  />
  <View
    style={
      Object {
        "flex": 1,
        "justifyContent": "flex-end",
        "paddingHorizontal": 20,
        "width": 750,
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "borderRadius": 3,
            "flexDirection": "row",
          },
          Object {
            "marginVertical": 10,
          },
          Object {
            "marginBottom": 10,
          },
          Array [
            Object {
              "fontFamily": "Hind-SemiBold",
              "fontSize": 16,
            },
          ],
          Object {
            "backgroundColor": "#42D689",
          },
        ]
      }
    >
      <View
        accessible={true}
        isTVSelectable={true}
        nativeBackgroundAndroid={
          Object {
            "attribute": "selectableItemBackground",
            "type": "ThemeAttrAndroid",
          }
        }
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Array [
            Object {
              "alignItems": "center",
              "flex": 1,
              "justifyContent": "center",
            },
            Object {
              "backgroundColor": "#42D689",
            },
            Object {
              "height": 50,
            },
            Object {
              "borderColor": "#42D689",
              "borderRadius": 3,
              "borderWidth": 2,
            },
          ]
        }
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "flex": 1,
              "flexDirection": "row",
              "justifyContent": "space-between",
            }
          }
        >
          <svg
            fill="none"
            height={12}
            style={Object {}}
            viewBox="0 0 10 12"
            width={10}
            xmlns="http://www.w3.org/2000/svg"
          >
            <line
              stroke="#FFFFFF"
              style={Object {}}
              x2="10"
              y1="11.5"
              y2="11.5"
            />
            <path
              d="M5 10L1.5359 5.5L8.4641 5.5L5 10Z"
              fill="#FFFFFF"
              style={Object {}}
            />
            <rect
              fill="#FFFFFF"
              height="7"
              style={Object {}}
              width="2"
              x="4"
            />
          </svg>
          <Text
            accessibilityLabel="shareCodeImage"
            style={
              Array [
                Object {
                  "fontFamily": "Hind-SemiBold",
                  "fontSize": 16,
                },
                Object {
                  "color": "#FFFFFF",
                },
                Object {
                  "paddingLeft": 5,
                  "paddingRight": 5,
                },
              ]
            }
          >
            shareCodeImage
          </Text>
        </View>
      </View>
    </View>
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "borderRadius": 3,
            "flexDirection": "row",
          },
          Object {
            "marginVertical": 10,
          },
          Object {
            "marginBottom": 10,
          },
          undefined,
          Object {
            "backgroundColor": "transparent",
          },
        ]
      }
    >
      <View
        accessible={true}
        isTVSelectable={true}
        nativeBackgroundAndroid={
          Object {
            "attribute": "selectableItemBackground",
            "type": "ThemeAttrAndroid",
          }
        }
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Array [
            Object {
              "alignItems": "center",
              "flex": 1,
              "justifyContent": "center",
            },
            Object {
              "backgroundColor": "transparent",
            },
            Object {
              "height": 50,
            },
            Object {
              "borderColor": "#42D689",
              "borderRadius": 3,
              "borderWidth": 2,
            },
          ]
        }
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "flex": 1,
              "flexDirection": "row",
              "justifyContent": "space-between",
            }
          }
        >
          <svg
            fill="none"
            height={16}
            style={Object {}}
            viewBox="0 0 20 16"
            width={20}
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M9.61121 5.66381C11.2646 5.66381 12.6101 7.00931 12.6101 8.66272C12.6101 10.3161 11.2646 11.6616 9.61121 11.6616C7.95779 11.6616 6.6123 10.3161 6.6123 8.66272C6.6123 7.00931 7.95779 5.66381 9.61121 5.66381ZM9.61121 4.32501C7.21543 4.32501 5.2735 6.26694 5.2735 8.66272C5.2735 11.0585 7.21543 13.0004 9.61121 13.0004C12.007 13.0004 13.9489 11.0585 13.9489 8.66272C13.9489 6.26694 12.0063 4.32501 9.61121 4.32501Z"
              fill="#42D689"
              style={Object {}}
            />
            <path
              d="M11.827 1.32541C12.0064 1.32541 12.175 1.4044 12.2895 1.5423L13.9664 3.55318C14.4028 4.07665 15.0434 4.37654 15.7249 4.37654H17.304C17.636 4.37654 17.9065 4.64697 17.9065 4.979V14.0721C17.9065 14.4041 17.636 14.6746 17.304 14.6746H1.92787C1.59585 14.6746 1.32541 14.4041 1.32541 14.0721V4.979C1.32541 4.64697 1.59585 4.37654 1.92787 4.37654H3.50699C4.18844 4.37654 4.82905 4.07665 5.2655 3.55318L6.94235 1.5423C7.05749 1.4044 7.22617 1.32541 7.4049 1.32541H11.827V1.32541ZM11.827 0H7.40423C6.8319 0 6.28968 0.253703 5.92352 0.693498L4.24734 2.70438C4.06393 2.92394 3.79282 3.05113 3.50699 3.05113H1.92787C0.862857 3.05113 0 3.91398 0 4.979V14.0721C0 15.1371 0.862857 16 1.92787 16H17.304C18.369 16 19.2319 15.1371 19.2319 14.0721V4.979C19.2319 3.91398 18.369 3.05113 17.304 3.05113H15.7249C15.439 3.05113 15.1679 2.92394 14.9845 2.70438L13.3077 0.693498C12.9415 0.253703 12.3993 0 11.827 0Z"
              fill="#42D689"
              style={Object {}}
            />
          </svg>
          <Text
            accessibilityLabel="scanCode"
            style={
              Array [
                Object {
                  "fontFamily": "Hind-SemiBold",
                  "fontSize": 16,
                },
                Object {
                  "color": "#42D689",
                },
                Object {
                  "paddingLeft": 5,
                  "paddingRight": 5,
                },
              ]
            }
          >
            scanCode
          </Text>
        </View>
      </View>
    </View>
  </View>
</View>
`;
